### DISCUSSION TITLE
How to keep queue consistence

### DISCUSSION THREAD

**Discussion by josecelano** [Nautilus-Cyberneering/git-queue#208] [General]
@da2ce7 have defined some cases we have to solve when queue messages are moved from now branch to other.

> With Git-Queue, the Git Repository is the synchronization authority.  I.e. We must not depend on any other external oracle for synchronization.
> 
> However, Git is a Decentralized Data Structure. So it may be forked at any point, where each fork is extended interdependently. For git, and it's commit histories, the point of resynchronization, is the 'merge', a special commit that can reference one, two, or many branched histories and bring them together.
> 
> ---
> 
> We have to cover the two cases:
> 1. An Active Git-Queue is branched, the active queue is cloned, and from that point, the job history diverges.
> 2. Multiple independent Git-Queues allocated on different branches, and never share a common starting ancestor, while sharing the same queue name.
> 
> We should detect if there is going to be a git-queue merge conflict.
> 
> To resolve this conflict we have two options.
> 1. Finish/Close/Delete any outstanding jobs, and then deallocate the queue on each branch before merging. Merge. Allocate a new Queue to replace the queue deallocated before the merge.
> 2. If the queues do not share a common ancestor, then we could select one to the be 'main', and only deallocate all the others.
> 
> As a note, if the queue was branched, I don't think that it is elegant to have a 'partly deallocated' queue.  We should deallocate all queues that share a common ancestor before merging branches of that queue.
> 
> We could also decide that it is best to deallocate all the conflicting queues, even if they do not share an ancestor.

I would add some more notes about this topic:

- Our current queue implementation is an ordered list of jobs, where you can add any number of new pending jobs (WIP) but jobs have to be completed in the order they were created.
- Our high-level queue abstraction could be an ordered list of jobs where all the incoming jobs can be in a pending a state.
- We have been using a 2-workflow (GitHub workflow) pattern to handle the jobs: a `work-allocator` which creates the new jobs and a `worker` which processes the jobs and marks them as finished.
- In a standard web application, for example, the simplest solution would be a SQL table with the list of jobs and a lock mechanism to avoid concurrency problems.
- We wanted to keep track not only of the final job state but also of all the job state changes. And we also wanted to put that information under git control, because we wanted to treat that information also as it was code, in order to get something like this in the Git history:

📝🈺: update_artwork: job.id.1 job.ref.ddc123312232baa3acb2342562fbc4535ccc234
📝▶️: update_artwork
📝⏸: update_artwork
📝✅: update_artwork: job.id.2: job.ref.232baa3acb2342562fbc4535ccc234ddc123312
source code commits ...
📝👔: update_artwork: job.id.2: job.ref.232baa3acb2342562fbc4535ccc234ddc123312
📝🈺: update_artwork: job.id.2: job.ref.232baa3acb2342562fbc4535ccc234ddc123312
📝✅: update_artwork: job.id.1: job.ref.1e31b549c630f806961a291b4e3d4a1471f37490
source code commits ...
📝👔: update_artwork: job.id.1: job.ref.1e31b549c630f806961a291b4e3d4a1471f37490
📝🈺: update_artwork: job.id.1: job.ref.1e31b549c630f806961a291b4e3d4a1471f37490
📝▶️: update_artwork

We wanted to track not only code changes but also the reason why that code was changed. Why? because some changes can be complex changes and they might be done automatically by other apps (GitHub bots, workflows, etcetera).

- Given those goals and given Git's nature it seemed that using the "event sourcing" pattern to store the Job state could be a good idea. So basically the `work-allocator` and `worker` call some commands in the GitHub action. The side effect is that some messages are recorded in the commit message of an empty commit. The queue can know the current state of a job by reading all the events (messages) related to that job.

![image](https://user-images.githubusercontent.com/58816/168307785-6e969761-c886-48b4-9f15-2e7d917f74d0.png)

This would be a simple and good solution if:

- **We have only one immutable event store**. Queue messages are stored in one Git branch. We could consider the "commits" like the records in the table. But even though commits are immutable we can not guarantee the consistency of the queue data structure. The queue data structure is an ordered list of empty commits, but commits can be changed outside the queue, for example with `git commit --amend`, `git rebase`, ...
- **We do not have queues in different branches** (or we handle those conflicts manually). Since Git is a decentralized data structure (as @da2ce7 said above), we could have another queue in a different branch with the same name. Right now we use the queue name as a global identifier.
- A special case of the previous one is a branch that contains a queue that is not finished. Right now, we do not have a way to know if a queue is still active or not but we planned to add some marks at the beginning and end of the queue to limit their scope in the Git history (@da2ce7 proposed allocation and deallocation command/messages).
- **We do not change the order of the commits**. Maybe we do not break the queue integrity but we can lose the order between the standard commits and the queue commits, which was important for us too.

In general, the problem is we cannot guarantee the queue consistency because the commits belonging to a queue can be changed while a merge, amend, rebase, or other Git operation is executed.

In the end, it seems we need an app on top of Git that checks the queue integrity after any Git command. It should fix or warn when there is a queue error, for example, a different queue in a feature branch has not been "deallocated/stopped/finished".

Some common problems could be:

- Unfinished queues.
- Messages with the wrong order. The order of the commits does not match the order of the messages.
- Mixing queues with the same name in different branches.

**Questions**

- Maybe the automatically generated commits should be also linked to the queue jobs. Maybe we could add a sign-off at the end of the commit message. In fact the [CommitMessageConventions](https://git.wiki.kernel.org/index.php/CommitMessageConventions) provides a particular case:

![image](https://user-images.githubusercontent.com/58816/168316620-aec18480-e3e0-476b-aca8-f84f06d3d4e9.png)

to [preserve the identity of a change after executing Git commands](https://gerrit-review.googlesource.com/Documentation/user-changeid.html). We could consider our jobs a given change. It's only an automatic change.

**Comment by josecelano** [comment]
I've been thinking of an alternative design. I have an idea but it's only a draft. I have to review whether it makes sense or not. These are the key points:

- Each Git commit it's a repo snapshot. Although we tend to see git as a linear history of changes it's more like a set of linked snapshots. You can even [replace](https://git-scm.com/docs/git-replace) those links.
- We want a reliable solution even if we use rebase or amend commands.
- The commit message conventions are a way to keep track of things on a higher level of abstraction (higher than commits).
- We need to keep information on why some commits happen. For example, you see a "new job" commit and after that the commits that do the job.
- Storing queue messages as an independent commit was a consequence of seeing git as a linear history.

And this could be a different option (pending to validate):

- We could store queue messages in independent orphan branches: "queue branches"
- When you allocate a new queue you create a new orphan branch for that queue. For example: `git checkout --orphan=update_artwork" target-branch-name`.
- The `work-allocator` creates and uses this special branch.
- The "queue branches" are "write models" to solve concurrency problems. Each commit is a queue message returned by a queue command.
- When the `worker` starts working on a job it will generate new commits while processing the job. Those jobs contain a full "read" model of the queue in the commit body. So we can know the queue status at the time that job commit was created.
- We could generate a `git-queue log` view with the list of "queue commits" and "job commits" in a linear way if we want it.
- We still use event sourcing to store the queue messages inside standard job commits. If fact, we can identify job commits because they have the "Git-queue: {JSON}" body field in the "Commit Message Conventions" section. We could use a different format to identify queue snapshots in the body if we do not want to follow "Commit Message Conventions".
- When a new job commit is generated we need to call the `git-queue snapshot` to get the current queue state that will be added to the commit body.
- The new git queue package would be a console command as we mentioned in our previous meetings.

Some of the `git-queue` commands could be:

- `git-queue allocate update_artwork target-branch-name`: To create the new queue branch.
- `git-queue new-job payload`: To create a new job. It creates a new empty commit in the queue branch.
- `git-queue snapshot`: it returns the current queue state to be added to the commit body.
- `git-queue start-job`.
- `git-queue finish-job`.
- `git-queue deallocate update_artwork: it deletes the remote queue branch.
- `git log`: git log but also showing queue messages between commit info. It uses information in the queue snapshot.

More notes:

- We can get into a conflict even before pushing job commits. All workers create new queue commits directly on the remote queue branch. We keep a linear history for the queue branch. We only allow fast forward merges.
- Commits messages could become big if queues contain a lot of messages. The only solution I see is start using files inside the repo instead of empty commits. That was one of the first ideas we had when we starting thinking about git-queue. We discarded it because we did not want to fix code with metadata. But using empty commits is like the same but "hidden" data. In the end, when we use DVC we are storing data inside the repo.

**Comment by da2ce7** [comment]
@josecelano Interesting.

I really like your prospective of seeing a git-commit as a snapshot: It is a much more representative mental model than what I previously had.

When considering comments as snapshots. That can be retaken; reference none, one or many parents; or even abstractly replaced entirely. Then we should be able to take any snapshot and _determine_ the state of the queue for that snapshot.

A snapshot could either have a consistent, or inconsistent git-queue status.  It is very possible that something goes wrong somewhere and the queue(s) associated with a particular snapshot are in a invalid state. We should detect any inconsistencies.

---

It begins to look like we have two fundamental options with git-queue's:

1. Branch Comments and Queues Fully Interlaced in the Branch History.
2. Extracting the Queue from the branch and putting it somewhere else (such as an orphan branch, per queue, as you suggest here).

We need to consider the tradeoffs between these two approaches.

In the case that we have some of the queue data stored in a different location to the git branch, we then need to consider how do we link these locations (by-directional linking?), and what happens when they become out-of-sync for any reason...

**Reply by josecelano** [reply]
hi @da2ce7 ,

> I really like your prospective of seeing a git-commit as a snapshot: It is a much more representative mental model than what I previously had.

The more I use Git the more I think I do not understand it very well so I started reading the Git Pro book. That point of view it's from [this chapter](https://git-scm.com/book/en/v2/Getting-Started-What-is-Git%3F). So credits go to @schacon. :-) 

> When considering comments as snapshots. That can be retaken; reference none, one or many parents; or even abstractly replaced entirely. Then we should be able to take any snapshot and determine the state of the queue for that snapshot.

The queue state changes in each commit. I think we somehow have to link a queue to a target branch. I mean, when you create (allocate) a queue your intention is to integrate some changes in a given branch. The `worker` can look for new jobs in that branch. When the `worker` finishes, it has to push atomically to the branch queue and the target branch. And as far as I know, that's not possible. And that leads me again to one of the first ideas we had for the implementation. If we want atomic push we have to use the same branch.

One of the first ideas we had was storing the queue data inside the repo, for example in a `.queue` directory at the root level (like `.git` and `.dvc`). We could have a file representing each queue. It could be a JSON object. We could also use one directory per queue and a file for each job (more git style). If a queue control commit (start, finished, ...) changes a job we have to guarantee that those files were not changed. In general, I think we have to make sure the queue state has not changed when we merge/push. If we store the queue state in a plain text json we could have no conflicts and if Git automatically merges them. One solution could be to store the queue data in a binary format, but that would not be user friendly. With this solution we won't have any size problem for the queue data. And it's a totally different approach. We do not store the queue messages using an event sourcing pattern. I would store the final jobs states. The history of queu changes would be the git history itself. You can get the queue history by parsing the queue data files after a git log. I suppose this is option 1 (Branch Comments and Queues Fully Interlaced in the Branch History) but using repo content instead of empty commits.

**Reply by josecelano** [reply]
hi @da2ce7 I think you will also find interesting this new CVS: [pijul](https://pijul.org/manual/why_pijul.html)

**Reply by da2ce7** [reply]
@josecelano  WOW! This is amazing, and written in rust to boot!

**Reply by josecelano** [reply]
I've added a new discussion in Nautilus org to follow up on this topic:
https://github.com/Nautilus-Cyberneering/.github/discussions/33
There is a [video on Youtube](https://www.youtube.com/watch?v=o0ooKVikV3c) from [Pierre Étienne Meunier](https://github.com/P-E-Meunier).

**Comment by josecelano** [comment]
I've found out that some projects are using custom git references to store simple information. inside the `.git` folder:

```
.git/refs/
├── heads
│   └── main
├── remotes
│   └── origin
│       └── HEAD
└── tags
```

- https://dvc.org/blog/dvc-2-0-release#lightweight-ml-experiments
- https://docs.graphite.dev/getting-started/why-use-stacked-changes#how-does-graphite-implement-stacked-changes
- https://gerrit-review.googlesource.com/Documentation/concept-refs-for-namespace.html

Another alternative to store data with Git is [annotated tags](https://stackoverflow.com/a/52125012/3012842).

**Reply by josecelano** [reply]
- [How Graphite stores branch metadata in references - TypeScript](https://github.com/withgraphite/graphite-cli/blob/6c80dff7b0339891af4c047c915518cf2bdab102/src/wrapper-classes/metadata_ref.ts)
- [Git as a key-value store - Python](https://github.com/edouardklein/gitkv)

**Reply by josecelano** [reply]
The Git Pro book also mentions [how you can use Git to store public GPG keys](https://github.com/Nautilus-Cyberneering/secure-git-guide/discussions/24).

**Comment by josecelano** [comment]
I've also found this interesting video:

[Using Git as a NoSql Database by Kenneth Truyers](https://www.youtube.com/watch?v=nPPlyjMlQ34&list=WL&index=2&t=723s)

They implemented this [NoSQl database using Git](https://github.com/YellowLineParking/Appy.GitDb).

I think they only implemented a "lock" mechanism per branch using C# `lock` function. So I suppose it only works when you have threads. Our workflows are secured in different machines.

**Reply by josecelano** [reply]
Some people say that video was copied from [this other one](https://www.youtube.com/watch?v=fjN4c4RWiV0).

At [this point](https://youtu.be/fjN4c4RWiV0?t=1465), he proposes to lock the Git ref file for race conditions. I suppose that only works for processes working with the same Git clone, which is not the case using different runners.

**Comment by josecelano** [comment]
I've been trying to define a new design for the queue based on storing jobs as files instead of empty commits. I hope it would be useful at least to discuss the pros and cons of this option.

### Proposal using files to store job current state

![queue-2 0-proposal-19-05-2022](https://user-images.githubusercontent.com/58816/169250511-98904f35-f804-45de-8fe3-2ae4d9e919ea.png)

### NOTES

- I think we could use the queue name hash (sha1) for the queue folder name, instead of a UUID. If we create the queue manually UUID should not be a problem, and it's an easy way to avoid conflicts merging queues. But if we also create the queue automatically it could lead to duplicate queues. The `work-allocator` could allocate the new queue twice, because it would create a new UUID for the second queue and there would not be any merge conflicts creating the two queues. I think in our current design we did not take into consideration the race condition even for the job creation commits. If I'm not wrong the same job could be created twice. It's less likely than create duplicate job commits because the execution time for the `work-allocator` should be constant.
- If we use the queue name or a hash of the queue name for the folder name, then we have to implement a "reallocation" command that would rename the queue to avoid the merge conflict.

The storage structure:

```s
.
└── .queue
    ├── 53b3ec1a908f8504e353f90ac43d680c7798199b
    │   ├── job-1.json
    │   ├── job-2.json
    │   ├── job-3.json
    │   └── queue.json
    └── .gitkeep
```

The `queue.json`:

```json
{
    "id": "53b3ec1a908f8504e353f90ac43d680c7798199b",
    "name": "update_artwork"
}
```

The `job-xx.json`:

```json
{
    "id": "c7172d00-7dc4-403a-b438-835facbc3b62",
    "index": "1",
    "state": "pending",
    "parent_queue_commit": "3a032b0992d7786b00a8822bbcbf192326160cf9",
    "queue-id": "53b3ec1a908f8504e353f90ac43d680c7798199b",
    "payload": "payload"
}
```

The commit body for job commits (commits created by the worker to perform the job) could contain links to the queue and job:

```
Git-queue-id: 53b3ec1a908f8504e353f90ac43d680c7798199b
Git-queue-job-id: c7172d00-7dc4-403a-b438-835facbc3b62
Sign-off: Bot <bot@email.com> 
```

**Comment by josecelano** [comment]
More comments on multi-job feature

We are implementing a [new feature to allow more than 1 pending job](https://github.com/Nautilus-Cyberneering/git-queue/issues/30).

Although it's a generic nice-to-have feature I think we can't use that feature in our use case: updating a consumer repo that is consuming a library repo as a git submodule.

We are currently using the git-queue in this [website project](https://github.com/Nautilus-Cyberneering/chinese-ideographs-website). The process works like this:

1. The `work-allocator` workflows detects changes in the git submodule ([the library](https://github.com/Nautilus-Cyberneering/chinese-ideographs)). It creates a new queue job in the `main` branch. The job payload is the current git submodule commit and the current upstream main HEAD.
2. The `worker` gets the job and processes the `diff` between both changes.

Right now there are some rules:

- We only allow on pending job.
- If two `work-allocator` workflows are executed in parallel only one of them will be able to create the job. The second job will be discarded.
- In that case, the job could be exactly the same (same commits for: `previous_ref` and `current_ref`) or it could be different in the upstream submodule has changed.
- We do not care about losing the second job because when the pending job is finished we will create a new job to synchronize the new changes. We are not mirroring atomic changes, we are just synchronizing the full state every time.
- We can't break the task down into smaller tasks because the submodule history could not be linear. We always need to compare both versions. If we were doing a backup this would be the same difference as a full or incremental backup. Since git commits are snapshot, when the origin snapshot changes we need to update the dependent snapshot in the consumer repo. And we need to calculate the differences and process them again.
- One on our original ideas was to  process each commit in the library. That would be a good way to limit the duration of the workflows execution since we have a limit.
- If the library git history were linear we could create new jobs to process the library changes. We could split the task into different jobs where each job processes a number of commits (or even only one).

Conclusion:

Even if we modify the queue to allow more than one pending job, we should not add more than one pending job in that case. Becuase we do not know if the intermediary steps (snapshots) would make things faster.

**Comment by josecelano** [comment]
Projects using git to store issues:

- https://github.com/MichaelMure/git-bug
- https://github.com/dspinellis/git-issue

**Comment by da2ce7** [comment]
@josecelano  Very Interesting, Good Find!  I knew about https://pagure.io/pagure that keeps all the comments in git-repo's however that is centralized.

[git-bug](https://github.com/MichaelMure/git-bug)'s internal model document: https://github.com/MichaelMure/git-bug/blob/master/doc/model.md is very interesting! :)

They also reference their use of a [conflict-free replicated data type (CRDT)](https://en.wikipedia.org/wiki/Conflict-free_replicated_data_type). That in itself is very interesting.

:)

**Comment by josecelano** [comment]
hi @da2ce7 @yeraydavidrodriguez I've been trying to summarize what I've learnt about using Git Objects and references to store your app data.

### Using Git as a key-value database

#### Where to store things?

With Git you can store data in two different ways:

- [Git objects](https://git-scm.com/book/en/v2/Git-Internals-Git-Objects)
- [References](https://git-scm.com/book/en/v2/Git-Internals-Git-References)

**Git objects**

Git internally uses a key-value database with only 4 [types of objects](https://git-scm.com/book/en/v2/Git-Internals-Git-Objects): blobs, trees, commits and annotated tags.

Each object is stored in the database and the way to reference the object is by using its sha1 (a checksum of the content).

For example, you can isert a new object in the database with:

```s
cd /tmp
mkdir test
git init
echo 'test content' | git hash-object -w --stdin
d670460b4b4aece5915caf5c68d12f560a9fe3e4
```

The last line is the sha1 of the new object. You can get the object with:

```s
git cat-file -p d670460b4b4aece5915caf5c68d12f560a9fe3e4
```

This database is immutable. You can only add new content.

**References**

Git has a dir `.git/refs` which contains files which are the references. Each file contains a sha1 of a database object.

If you want to get the content from the previous example but do not want to use the sha1 you could create a reference like this:

```
git update-ref refs/my-objects/object-1 d670460b4b4aece5915caf5c68d12f560a9fe3e4
```

The reference is a pointer or alias for the object hash you have inserted in the database before. And now you can use the reference to get the content:

```s
$ git cat-file -p refs/my-objects/object-1
test content
```

This database is a mutable database. You can change references to point to different Git objects.

You can create new references in any fork of the repo and push them to any remote repo.

There are some special references handled by Git:

```s
$ tree -al .git/refs/
.git/refs/
├── heads
├── remotes
├── my-objects
│   └── object-1
└── tags
```

Branches are only references pointing to the latest commit object hash in a sequence of commits. When you run a `git push` command Git tries to update a reference in a remote repo.

#### How to store things?

Git provides only those basic low-level databases to store things: objects and references. You can use them in different ways. We have seen at least two different models:

**Store only the latest state**

You can store your data inside blob objects. When you want to update the version of your object you can store a new object. You can use a reference to retrieve the blob object. You also have to avoid the [Git Gargable Collector](https://git-scm.com/docs/git-gc) to remove the object. If the blob object is not referenced anywhere it could be deleted.

![image](https://user-images.githubusercontent.com/58816/170452306-a5ba38dc-ecf9-4e01-bc01-4d5acd00baae.png)

**Pros**

- Simple

**Cons**

- You can’t checkout the reference because the has does not point to a commit.
- You can’t follow changes on the object because each new object version is not linked to the previous one. It’s like a standard database. If you want to keep a history of your changes you can do it by yourself or use the next model.
- It generates a lot of references.

**Store state change history (commits)**

The previous solution allows you to store objects like a key-value database. But you can take full advantage of Git by using the other objects available. When you update a blob object you could create a commit.

This model would be the same as using an orphan branch for each object.

```s
cd /tmp
mkdir test
cd test
git init
git checkout --orphan my-objects-object-1
echo "test content" > object-1.txt
git add object-1.txt 
git commit -m "add object-1"
git checkout my-objects-object-1 && cat object-1.txt 
```

You can get the object with:

```s
git checkout my-objects-object-1 && cat object-1.txt 
```

```s
$ tree -al .git/refs/
.git/refs/
├── heads
│   └── my-objects-object-1
└── tags
```

![image](https://user-images.githubusercontent.com/58816/170452403-53f192ca-a5e0-4e9e-99e0-4b34cc97b5c7.png)

**Pros**

- You can add metadata for every database operation in the commit message.
- You can easily get the data using the common Git porcelain commands.
- You have the history of changes for your objects.

**Cons**

- It's More complex than the previous solution.
- It generates a lot of references.

#### How to solve race conditions

We have seen two possible models to use Git to store your data as a key-value database. But does this DB  implementation offer you a way to handle race conditions?

At some point, you are going to have two processes reading the same object and trying to update it at the same time. One of them is going to update from a previous value.

We can create an example where we have a “table” with counters.

We insert the object with the counter starting at 0.

```s
cd /tmp
mkdir my-counters
cd my-counters/
git init
git checkout -b my-counters-counter-1
echo "0" > counter-1.txt
git add -A
git commit -m "initialize counter-1"
```

Independent processes could checkout the repo and increment the counter. After cloning the repo you have an old version of the data because other processes could have cloned and updated the counter.

![image](https://user-images.githubusercontent.com/58816/170452557-c487ec12-9e4a-48d7-9856-ef5eb1cf02d6.png)

**Optimistic Concurrency Control**

We normally have two options to fix that problem with normal databases, you can either lock the record when you want to modify it (pessimistic locking) or try to modify it always and make the update fail if the record has changed (optimistic locking).

Git only allows us to use the “optimistic approach”. When you try to “push” your object version by updating the reference in the origin repo you will get an error if the reference (branch) was already changed.

![image](https://user-images.githubusercontent.com/58816/170452750-4d5332b9-4b74-47c0-a983-7732d0834e61.png)

Increment the counter in the process 1:

```s
cd /tmp
git clone /tmp/my-counters my-counters-process-1
cd /tmp/my-counters-process-1/
git checkout my-counters-counter-1
echo "1" > counter-1.txt 
git add counter-1.txt 
git commit -m "increment counter to 1"
```

You can do the same with a clone for process 2. Then you can push the changes from process 1. If you try to push the counter in the origin repo you could have this error:

```s
git push origin
…
 ! [remote rejected] counter-1 -> counter-1 (branch is currently checked out)
error: failed to push some refs to '/tmp/my-counters'
```

You only need to checkout a different branch in the origin repo.

After pushing from the process 1 fork and try to push from the second one you will see the Git error:

```s
On branch counter-1
Your branch is ahead of 'origin/counter-1' by 1 commit.
  (use "git push" to publish your local commits)

nothing to commit, working tree clean
$ git push
To /tmp/my-counters
 ! [rejected]        counter-1 -> counter-1 (fetch first)
error: failed to push some refs to '/tmp/my-counters'
hint: Updates were rejected because the remote contains work that you do
hint: not have locally. This is usually caused by another repository pushing
hint: to the same ref. You may want to first integrate the remote changes
hint: (e.g., 'git pull ...') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.
```

You cannot update the reference because the previous commits created by process 1 would be lost.

Git does not allow you to update the object if someone else has already changed it. So basically if you use Git to store your objects the only mechanism to avoid race conditions is optimistic locking. 

This way of using Git is like having an SQL table where you have version numbers for record updates by default. That means every time you read a record you get the version number. When you update the record you check that the version number is still the same. The only difference compared with the normal SQL optimistic lock approach is that you always store the new version of the object, but you do not update the reference (pointer) to it. So you will continue retrieving the previous version.

#### How to design your unit of work

Since you only can control concurrency at the object level (references) you have to make sure that you [design your aggregates](https://www.dddcommunity.org/library/vernon_2011/) in a way that each aggregate is a different reference (or branch if you use the second model).

**What does that mean?**

In the same example, you can have strict domain rules between counters. For example:

- You cannot have more than a certain number of counters.
- All counters cannot count more than 100.

In those cases, you should have to define an object like a "pool of counters" and store all the values in the same object (branch).

I recommend reading [these articles ](https://www.dddcommunity.org/library/vernon_2011/) to know the trade-offs of aggregate design.

#### How to store the Git Queue

We have not defined yet the final requirements for Git Queue 2.0. Some open questions are:

- Are we going to use only one queue or two queues? One for job submission and one for job completion.
- Are we going to allow more types of dependencies between jobs? For now, we only allow execute jobs in the order they were created. We could allow users to define job dependencies.
- Do we want some parameters to be configurable? For example maximum number of pending jobs.

If we do not have any invariant between jobs we could store each job in a different orphan branch. That way we can reduce conflicts and have a better performance. On the contrary, if we have some invariants between jobs then we have to store the whole queue in the same branch, so we make sure we do not have inconsistencies. We could decide the storage method depending on the invariants declared in the queue configuration at runtime. For example, the user could declare the queue as a queue whose “jobs do not have any dependency”.

#### Atomic push

Job consumers (workers) have to process the jobs. In our case (update Git submodule), the result is a set of commits you want to merge into the target branch.

In our case, we want to update the queue and merge the “job commits” atomically, otherwise, you can have inconsistencies between the job done and the queue state. If you push the job commits but you cannot push the job update, the worker will try to process again the same job. On the contrary, you could push the job update to the queue branch, but you can have an error pushing the job commits.

The queue update and the new commits have to be pushed atomically to their branches. Fortunately, Git has a push option “--atomic” that does exactly that: "Either all refs are updated, or on error, no refs are updated”.

So you could do something like:

```
git push --atomic origin main refs/queues/update-upwork
```

More info about Atomic pushes:

- https://github.blog/2015-04-30-git-2-4-atomic-pushes-push-to-deploy-and-more/
- https://git-scm.com/docs/git-push#Documentation/git-push.txt---no-atomic

#### Projects using Git objects and references to store data

- https://graphite.dev/
- https://dvc.org/doc/user-guide/experiment-management/experiments-overview#how-does-dvc-track-experiments

#### Talks

- [Using Git as a NoSql Database by Kenneth Truyers](https://www.youtube.com/watch?v=nPPlyjMlQ34)

#### Articles

- [How Git truly works](https://towardsdatascience.com/how-git-truly-works-cd9c375966f6)
- [The Biggest Misconception About Git](https://medium.com/@gohberg/the-biggest-misconception-about-git-b2f87d97ed52)

**Reply by da2ce7** [reply]
@josecelano Answering the questions we had about if `git push` would reject overwriting a ref outside of the branch folder.

The good news is that it behaves sensibly:

1. REJECT updating any remote reference that refers to an object not in the local git object store.

2. REJECT updating any remote reference that refers to a non-comment-object.

3. REJECT updating any remote reference that changes a comment-object into referring a non comment-object.

4. REJECT updating any remote reference that isn't a direct parent of the new-comment-object. (ie. non-fast-forward push)

ie. It only accepts fast forward comment reference updates.

**Reply by josecelano** [reply]
OK @da2ce7 cool!

So I suppose with point 1 we cover model 1 (blobs and references) and with point 4 we cover model 2 (orphan branches).

Did you test it with custom references?

By the way, the [Graphite cli](https://github.com/withgraphite/graphite-cli) is in TypeScript and it has a lot of code related to Git.
